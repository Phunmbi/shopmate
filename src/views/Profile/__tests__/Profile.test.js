import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import ProfileContainer,{ Profile } from '../index';

describe('<ConnectedHomePage />', () => {
  const props = {
    retrieveCart: jest.fn(),
    removeFromCart: jest.fn(),
    updateUser: jest.fn(),
    updateCart: jest.fn(),
    getUser: jest.fn(),
    getShippingRegions: jest.fn(),
    cart: [
      {name: 'adeniyi', attributes: "Blue, S", item_id: 1, quantity: 3},
      {name: 'victor', attributes: "Purple, S", item_id: 2, quantity: 4}],
    history: {
      push: jest.fn()
    },
    productDetails: {
      product_id: 1,
      name: 'nigerian',
      subtotal: 455,
      thumbnail: 'thumbnail.png',
      image_2: 'image2.png',
      image: 'image.png'
    },
    reviews: [{name: 'named', review: 'Terrible product', rating: 4}],
    cartLoading: false,
    allProducts: [],
    shippingRegions: [
      {shipping_region: "Canada/Us", shipping_region_id: 2},
      {shipping_region: "Rest of the World", shipping_region_id: 1}
    ]
  };

  const initialState = {
    products: {
      products: [
        {name: 'altica', product_id: 12, discounted_price: "0.00"},
        {name: 'italia', product_id: 19},
        {name: 'french', product_id: 17},
        {name: 'nigerian', product_id: 15},
        {name: 'christmas cap', product_id: 14}
      ],
      singleProductDetails: {
        product_id: 1,
        name: 'nigerian',
        subtotal: 455,
        thumbnail: 'thumbnail.png',
        image_2: 'image2.png',
        image: 'image.png'
      },
      count: 5,
      loading: false,
      reviews: [{name: 'named', review: 'Terrible product', rating: 4}],
      reviewsLoading: false,
      singleProductLoading: false,
    },
    auth: {
      loading: false,
      user: {
        name: 'Name',
      }
    },
    checkout: {
      shippingRegions: [{shipping_region: "Canada/Us", shipping_region_id: 2},
        {shipping_region: "Rest of the World", shipping_region_id: 1}]
    },
    shoppingCart: {
      cart: [{name: 'adeniyi', attributes: "Blue, S", item_id: 1, quantity: 3},
        {name: 'victor', attributes: "Purple, S", item_id: 2, quantity: 4}],
      cart_Id: 4,
      loadingShoppingCart: false
    },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  let wrapper;

  beforeEach(() => {
    localStorage.setItem("name", "named");
    wrapper = mount(<Profile {...props} store={store}/>);
  });

  it('renders appropriately', () => {
    localStorage.setItem("name", "named");
    const wrapped = mount(
      <Provider store={store}>
        <MemoryRouter key="keyed">
          <ProfileContainer {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapped).toMatchSnapshot();
  });

  it('should successfully render signup modal and close it', () => {
    wrapper.setState({displayModal: true, openModal: 'signup'});

    wrapper.find('SignUp').find('.modal-overlay').simulate('click');

    expect(wrapper.find('SignUp').exists()).toEqual(true);
  });

  it('should successfully render signup modal and switch to signin modal', () => {
    wrapper.setState({displayModal: true, openModal: 'signup'});

    wrapper.find('SignUp').find('.additional-options__alternate').find('p').simulate('click');

    expect(wrapper.find('SignIn').exists()).toEqual(true);
  });

  it('should successfully render signin modal', () => {
    wrapper.setState({displayModal: true, openModal: 'signin'});

    expect(wrapper.find('SignIn').exists()).toEqual(true);
  });

  it('should successfully render checkout modal', () => {
    wrapper.setState({displayModal: true, openModal: 'checkoutCart'});

    expect(wrapper.find('CheckoutCart').exists()).toEqual(true);
  });

  it('should simulate updateCart', () => {
    //Set in test local storage mocks
    localStorage.setItem('1', JSON.stringify({quantity: 3}));
    localStorage.setItem('2', JSON.stringify({quantity: 4}));
    localStorage.setItem('cart_id', '3');

    wrapper.setState({displayModal: true, openModal: 'checkoutCart'});
    wrapper.find('CheckoutCart').setState({ changingQuantity: true , 1: {quantity: 3}, 2: {quantity: 4}});
    const inst = wrapper.instance();
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(1).simulate('click');

    wrapper.find('.active-button').simulate('click');

    inst.handleUpdateCart = jest.fn();
    wrapper.find('.active-button').simulate('click');

    expect(wrapper.find('CheckoutCart').exists()).toEqual(true);
  });

  it('should simulate updateCart and simulate checkout while logged out', () => {
    //Set in test local storage mocks
    localStorage.setItem('1', JSON.stringify({quantity: 3}));
    localStorage.setItem('2', JSON.stringify({quantity: 4}));
    localStorage.setItem('cart_id', '3');

    wrapper.find('ProfileForm').setState({
      values: {
        'Address 1': '1 aminu street',
        'Address 2': 'address 2',
        City: 'Lagos',
        Region: 'South West',
        Country: 'Nigeria',
        'Postal Code': '22321',
        'Shipping Region': 'Rest of the World',
      },
      errors: {
        'Address 1': 'Error out here',
        'Address 2': 'Error out here',
        City: 'Error out here',
        Region: 'Error out here',
        Country: 'Error out here',
        'Postal Code': 'Error out here',
        'Shipping Region': 'Error out here'
      }
    });

    const inst = wrapper.instance();

    wrapper.setState({displayModal: true, openModal: 'checkoutCart'});
    wrapper.find('CheckoutCart').setState({ changingQuantity: true , 1: {quantity: 3}, 2: {quantity: 4}});
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(1).simulate('click');

    wrapper.find('.checkout-button').find('button').at(2).simulate('click');

    inst.startCheckout = jest.fn();

    expect(wrapper.find('CheckoutCart').exists()).toEqual(false);
  });

  it('should simulate updateCart and simulate checkout, while logged in', () => {
    //Set in test local storage mocks
    localStorage.setItem('1', JSON.stringify({quantity: 3}));
    localStorage.setItem('2', JSON.stringify({quantity: 4}));
    localStorage.setItem('isAuthenticated', 'true');
    localStorage.setItem('cart_id', '3');

    wrapper.setProps({user: {name: 'done'}});

    const inst = wrapper.instance();

    wrapper.setState({displayModal: true, openModal: 'checkoutCart'});
    wrapper.find('CheckoutCart').setState({ changingQuantity: true , 1: {quantity: 3}, 2: {quantity: 4}});
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(0).simulate('click');
    wrapper.find('.checkout-container__quantity').find('img').at(1).simulate('click');

    wrapper.find('.checkout-button').find('button').at(2).simulate('click');

    inst.startCheckout = jest.fn();
    wrapper.find('.checkout-button').find('button').at(2).simulate('click');

    expect(wrapper.find('CheckoutCart').exists()).toEqual(true);
  });

  it('should simulate handleSelectRegion in shipping region', () => {
    wrapper.find('ProfileForm').find('.profile-form__shipping').find('select').simulate('change');
    expect(wrapper.find('ProfileForm').exists()).toEqual(true);
  });

  it('should simulate handleChange in shipping region', () => {
    wrapper.find('ProfileForm').find('form').find('input').at(1).simulate('change');
    expect(wrapper.find('ProfileForm').exists()).toEqual(true);
  });

  it('should simulate handleChange in shipping region with errors in validation', () => {
    wrapper.find('ProfileForm').find('form').find('input').at(1).simulate('change',  {target: {title: 'Address 1', value: 'aminu'}});
    wrapper.find('ProfileForm').find('form').find('input').at(2).simulate('change',  {target: {title: 'Address 2', value: ' '}});
    wrapper.find('ProfileForm').find('form').find('input').at(3).simulate('change',  {target: {title: 'City', value: 'Lagos'}});
    wrapper.find('ProfileForm').find('form').find('input').at(4).simulate('change',  {target: {title: 'Region', value: 'South West'}});
    wrapper.find('ProfileForm').find('form').find('input').at(5).simulate('change',  {target: {title: 'Country', value: 'Nigeria'}});
    wrapper.find('ProfileForm').find('form').find('input').at(6).simulate('change',  {target: {title: 'Postal Code', value: '434'}});
    expect(wrapper.find('ProfileForm').exists()).toEqual(true);
  });

  it('should simulate handleSubmit in Profile Form', () => {
    wrapper.find('ProfileForm').setState({
      values: {
        'Address 1': ' 1 aminu street',
        'Address 2': 'address 2',
        City: 'Lagos',
        Region: 'South West',
        Country: 'Nigeria',
        'Postal Code': '22321',
        'Shipping Region': 'Rest of the World',
      },
      errors: {}
    });
    wrapper.find('ProfileForm').find('form').simulate('submit');
    expect(wrapper.find('ProfileForm').exists()).toEqual(true);
  });
});
