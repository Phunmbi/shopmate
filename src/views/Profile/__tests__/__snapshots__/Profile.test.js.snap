// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectedHomePage /> renders appropriately 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    key="keyed"
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "aa05a9",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "aa05a9",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Profile)
        allProducts={Array []}
        cart={
          Array [
            Object {
              "attributes": "Blue, S",
              "item_id": 1,
              "name": "adeniyi",
              "quantity": 3,
            },
            Object {
              "attributes": "Purple, S",
              "item_id": 2,
              "name": "victor",
              "quantity": 4,
            },
          ]
        }
        cartLoading={false}
        getShippingRegions={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        getUser={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        history={
          Object {
            "push": [MockFunction],
          }
        }
        productDetails={
          Object {
            "image": "image.png",
            "image_2": "image2.png",
            "name": "nigerian",
            "product_id": 1,
            "subtotal": 455,
            "thumbnail": "thumbnail.png",
          }
        }
        removeFromCart={[MockFunction]}
        retrieveCart={
          [MockFunction] {
            "calls": Array [
              Array [
                null,
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        reviews={
          Array [
            Object {
              "name": "named",
              "rating": 4,
              "review": "Terrible product",
            },
          ]
        }
        shippingRegions={
          Array [
            Object {
              "shipping_region": "Canada/Us",
              "shipping_region_id": 2,
            },
            Object {
              "shipping_region": "Rest of the World",
              "shipping_region_id": 1,
            },
          ]
        }
        updateCart={[MockFunction]}
        updateUser={[MockFunction]}
      >
        <Profile
          allProducts={
            Array [
              Object {
                "discounted_price": "0.00",
                "name": "altica",
                "product_id": 12,
              },
              Object {
                "name": "italia",
                "product_id": 19,
              },
              Object {
                "name": "french",
                "product_id": 17,
              },
              Object {
                "name": "nigerian",
                "product_id": 15,
              },
              Object {
                "name": "christmas cap",
                "product_id": 14,
              },
            ]
          }
          authLoading={false}
          cart={
            Array [
              Object {
                "attributes": "Blue, S",
                "item_id": 1,
                "name": "adeniyi",
                "quantity": 3,
              },
              Object {
                "attributes": "Purple, S",
                "item_id": 2,
                "name": "victor",
                "quantity": 4,
              },
            ]
          }
          cartLoading={false}
          cart_Id={4}
          getShippingRegions={[Function]}
          getUser={[Function]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          productDetails={
            Object {
              "image": "image.png",
              "image_2": "image2.png",
              "name": "nigerian",
              "product_id": 1,
              "subtotal": 455,
              "thumbnail": "thumbnail.png",
            }
          }
          productsCount={5}
          productsLoading={false}
          productsResponse={
            Object {
              "count": 5,
              "loading": false,
              "products": Array [
                Object {
                  "discounted_price": "0.00",
                  "name": "altica",
                  "product_id": 12,
                },
                Object {
                  "name": "italia",
                  "product_id": 19,
                },
                Object {
                  "name": "french",
                  "product_id": 17,
                },
                Object {
                  "name": "nigerian",
                  "product_id": 15,
                },
                Object {
                  "name": "christmas cap",
                  "product_id": 14,
                },
              ],
              "reviews": Array [
                Object {
                  "name": "named",
                  "rating": 4,
                  "review": "Terrible product",
                },
              ],
              "reviewsLoading": false,
              "singleProductDetails": Object {
                "image": "image.png",
                "image_2": "image2.png",
                "name": "nigerian",
                "product_id": 1,
                "subtotal": 455,
                "thumbnail": "thumbnail.png",
              },
              "singleProductLoading": false,
            }
          }
          removeFromCart={[Function]}
          retrieveCart={[Function]}
          reviews={
            Array [
              Object {
                "name": "named",
                "rating": 4,
                "review": "Terrible product",
              },
            ]
          }
          shippingRegions={
            Array [
              Object {
                "shipping_region": "Canada/Us",
                "shipping_region_id": 2,
              },
              Object {
                "shipping_region": "Rest of the World",
                "shipping_region_id": 1,
              },
            ]
          }
          updateCart={[Function]}
          updateUser={[Function]}
          user={
            Object {
              "name": "Name",
            }
          }
        >
          <div
            className="homepage"
          >
            <header
              className="navbar"
            >
              <NavbarHome
                bagTotal={[Function]}
                cartCount={2}
                handleDisplayModal={[Function]}
                retrieveCart={[Function]}
              >
                <div
                  className="navbar-home"
                >
                  <div
                    className="navbar-home__main"
                  >
                    <div
                      className="navbar-home__action .Text-Style-2"
                    >
                      <h3
                        className="navbar-home__action1"
                      >
                        Hi! 
                      </h3>
                      <h3
                        className="navbar-home__action2"
                        onClick={[Function]}
                      >
                        Sign in
                      </h3>
                      <h3
                        className="navbar-home__action1"
                      >
                        or 
                      </h3>
                      <h3
                        className="navbar-home__action2"
                        onClick={[Function]}
                      >
                        Register
                      </h3>
                    </div>
                    <div
                      className="navbar-home__cart"
                      onClick={[Function]}
                    >
                      <Cart
                        cartCount={2}
                      >
                        <div
                          className="cart-main"
                        >
                          <img
                            alt="Checkout Bag"
                            src="cart-white.svg"
                          />
                          <div
                            className="cart-main__count"
                          >
                            2
                          </div>
                        </div>
                      </Cart>
                      <div>
                        Your Bag: Â£
                        NaN
                      </div>
                    </div>
                  </div>
                </div>
              </NavbarHome>
              <NavBarCollapse
                history={
                  Object {
                    "push": [MockFunction],
                  }
                }
              >
                <div
                  className="navbar-collapse"
                >
                  <div
                    className="navbar-collapse__container"
                  >
                    <h3
                      onClick={[Function]}
                    >
                      SHOPMATE
                    </h3>
                  </div>
                </div>
              </NavBarCollapse>
            </header>
            <div
              className="profile"
            >
              <ProfileForm
                getShippingRegions={[Function]}
                getUser={[Function]}
                history={
                  Object {
                    "push": [MockFunction],
                  }
                }
                shippingRegions={
                  Array [
                    Object {
                      "shipping_region": "Canada/Us",
                      "shipping_region_id": 2,
                    },
                    Object {
                      "shipping_region": "Rest of the World",
                      "shipping_region_id": 1,
                    },
                  ]
                }
                updateUser={[Function]}
                userDetails={
                  Object {
                    "name": "Name",
                  }
                }
              >
                <div
                  className="profile-form"
                >
                  <form
                    onSubmit={[Function]}
                  >
                    <input
                      disabled={true}
                      placeholder="Name"
                      type="text"
                      value="named"
                    />
                    <input
                      onChange={[Function]}
                      placeholder="Address 1"
                      title="Address 1"
                      type="text"
                    />
                    <input
                      onChange={[Function]}
                      placeholder="Address 2"
                      title="Address 2"
                      type="text"
                    />
                    <div
                      className="profile-form__location"
                    >
                      <input
                        onChange={[Function]}
                        placeholder="City"
                        title="City"
                        type="text"
                      />
                      <input
                        onChange={[Function]}
                        placeholder="Region"
                        title="Region"
                        type="text"
                      />
                      <input
                        onChange={[Function]}
                        placeholder="Country"
                        title="Country"
                        type="text"
                      />
                    </div>
                    <div
                      className="profile-form__shipping"
                    >
                      <input
                        onChange={[Function]}
                        placeholder="Postal Code"
                        title="Postal Code"
                        type="text"
                      />
                      <select
                        name="shipping-region"
                        onChange={[Function]}
                      >
                        <option
                          key="2"
                          label="Canada/Us"
                          value={2}
                        >
                          Canada/Us
                        </option>
                        <option
                          key="1"
                          label="Rest of the World"
                          value={1}
                        >
                          Rest of the World
                        </option>
                      </select>
                    </div>
                    <input
                      type="submit"
                      value="Submit"
                    />
                  </form>
                </div>
              </ProfileForm>
            </div>
            <footer
              className="footer"
            >
              <Footer>
                <div
                  className="footer-main"
                >
                  <div
                    className="footer-container"
                  >
                    <div
                      className="footer-container__categories"
                    >
                      <h3>
                        Women
                      </h3>
                      <h3>
                        Men
                      </h3>
                      <h3>
                        Kids
                      </h3>
                      <h3>
                        Shoes
                      </h3>
                      <h3>
                        Brands
                      </h3>
                    </div>
                    <div
                      className="footer-container__social"
                    >
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="instagram"
                          src="instagramIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="pinterest"
                          src="pinterestIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="twitter"
                          src="twitterIcon.svg"
                        />
                      </div>
                      <div
                        className="footer-container__socialIcons"
                      >
                        <img
                          alt="facebook"
                          src="facebookIcon.svg"
                        />
                      </div>
                    </div>
                    <div
                      className="footer-container__information"
                    >
                      Â© shopmate Ltd â¢ Contact â¢ Privacy Policy
                    </div>
                  </div>
                </div>
              </Footer>
            </footer>
          </div>
        </Profile>
      </Connect(Profile)>
    </Router>
  </MemoryRouter>
</Provider>
`;
